From ee6f87ee5bf3bd7c9dc382b84560555b3e92840b Mon Sep 17 00:00:00 2001
From: "Brett T. Warden" <brett.t.warden@intel.com>
Date: Thu, 2 Aug 2018 10:10:32 -0700
Subject: [PATCH] Add stateless support in /var/lib/xastir

Add search functionality to try, in order:
* Environment variable XASTIR_DATA_BASE
* Path /var/lib/xastir
* Path /usr/share/xastir

Update scripts that fetch public databases to store them under
/var/lib/xastir
---
 configure.ac              |  3 ++-
 scripts/get-BOMdata       |  3 ++-
 scripts/get-NWSdata       |  3 ++-
 scripts/get-fcc-rac.pl    |  3 ++-
 scripts/get-gnis          |  7 ++++---
 scripts/get-pop           | 12 +++++++-----
 scripts/xastir-fixcfg.sh  |  2 +-
 scripts/xastir-migrate.sh |  2 +-
 src/xa_config.c           | 33 ++++++++++++++++++++++++++-------
 9 files changed, 47 insertions(+), 21 deletions(-)

diff --git a/configure.ac b/configure.ac
index 1ef8e45..b137da0 100644
--- a/configure.ac
+++ b/configure.ac
@@ -688,7 +688,8 @@ if test "x${datadir}" = "x"; then
 else 
   CPPFLAGS="$CPPFLAGS -DXASTIR_DATA_BASE=\\\"${datadir}/xastir\\\""
 fi 
- 
+# Add stateless support for additional data dir
+CPPFLAGS="$CPPFLAGS -DXASTIR_STATELESS_BASE=\\\"/var/lib/xastir\\\""
 
 
 # Add a special option that makes Cygwin link much faster (from
diff --git a/scripts/get-BOMdata b/scripts/get-BOMdata
index 540046d..e5c66b8 100755
--- a/scripts/get-BOMdata
+++ b/scripts/get-BOMdata
@@ -42,7 +42,8 @@ FILE7="LGA08aAust"                   # Local Government Area Boundaries
 
 x=`dirname $0`
 . $x/values
-cd ${prefix}/share/xastir/Counties
+mkdir -p /var/lib/xastir/Counties
+cd /var/lib/xastir/Counties
 
 
 # Remove any old zip files hanging around in this directory
diff --git a/scripts/get-NWSdata b/scripts/get-NWSdata
index 7720afa..d2d98c7 100755
--- a/scripts/get-NWSdata
+++ b/scripts/get-NWSdata
@@ -56,7 +56,8 @@ FILE7="c_11au16"    # AWIPS County Libraries (under "States, Provinces & Countie
 
 x=`dirname $0`
 . $x/values
-cd ${prefix}/share/xastir/Counties
+mkdir -p /var/lib/xastir/Counties
+cd /var/lib/xastir/Counties
 
 
 # Remove any old zip files hanging around in this directory
diff --git a/scripts/get-fcc-rac.pl b/scripts/get-fcc-rac.pl
index 544d563..d838c4f 100755
--- a/scripts/get-fcc-rac.pl
+++ b/scripts/get-fcc-rac.pl
@@ -27,7 +27,7 @@ use File::Basename;
 $dirname=dirname($0);
 require ($dirname."/values.pl");
 
-my $XASTIR_BASE="${prefix}/share/xastir";
+my $XASTIR_BASE="/var/lib/xastir";
 
 
 # This script uses temporary storage space in /var/tmp to do its work.
@@ -59,6 +59,7 @@ if (-e $file && -r $file && -f $file) {
   print STDERR "*** Installing the RAC database ***\n";
   print STDERR "***********************************\n";
   `unzip $file $file2`;
+  `mkdir -p $XASTIR_BASE/fcc`;
   `mv $file2 $XASTIR_BASE/fcc/AMACALL.LST`;
 }
 
diff --git a/scripts/get-gnis b/scripts/get-gnis
index 2e21384..0937c66 100755
--- a/scripts/get-gnis
+++ b/scripts/get-gnis
@@ -57,10 +57,11 @@ do
 
     if ( [ -f ${MYSTATE}${SUFFIX}.txt ] )
     then 
-        printf "File successfully downloaded. Moving to ${prefix}/share/xastir/GNIS\n" 
-        sudo mv ${MYSTATE}${SUFFIX}.txt ${prefix}/share/xastir/GNIS/${MYSTATE}.gnis
+        printf "File successfully downloaded. Moving to /var/lib/xastir/GNIS\n" 
+        sudo mkdir -p /var/lib/xastir/GNIS
+        sudo mv ${MYSTATE}${SUFFIX}.txt /var/lib/xastir/GNIS/${MYSTATE}.gnis
 	if [ ${MYSTATE} = "AK" -o ${MYSTATE} = "HI" ]; then
-		sudo recode utf16..utf8 ${prefix}/share/xastir/GNIS/${MYSTATE}.gnis
+		sudo recode utf16..utf8 /var/lib/xastir/GNIS/${MYSTATE}.gnis
 	fi
     else 
         printf "File for %s not successfully downloaded.\n" ${MYSTATE}
diff --git a/scripts/get-pop b/scripts/get-pop
index e0641f4..e98f0ef 100755
--- a/scripts/get-pop
+++ b/scripts/get-pop
@@ -61,14 +61,16 @@ do
     fi
 
     if ( [ -f ${MYSTATE}${SUFFIX}.txt ] ) then 
-        printf "File successfully downloaded. Moving to ${prefix}/share/xastir/GNIS\n" 
-        sudo mv ${MYSTATE}${SUFFIX}.txt ${prefix}/share/xastir/GNIS/${MYSTATE}.pop
+        printf "File successfully downloaded. Moving to /var/lib/xastir/GNIS\n" 
+        sudo mkdir -p /var/lib/xastir/GNIS
+        sudo mv ${MYSTATE}${SUFFIX}.txt /var/lib/xastir/GNIS/${MYSTATE}.pop
 	if [ ${MYSTATE} = "AK" -o ${MYSTATE} = "HI" ]; then
-		sudo recode utf16..utf8 ${prefix}/share/xastir/GNIS/${MYSTATE}.pop
+		sudo recode utf16..utf8 /var/lib/xastir/GNIS/${MYSTATE}.pop
 	fi
     elif ( [ -f ${MYSTATE}${SUFFIX} ] ) then 
-        printf "File successfully downloaded. Moving to ${prefix}/share/xastir/GNIS\n" 
-        sudo mv ${MYSTATE}${SUFFIX} ${prefix}/share/xastir/GNIS/${MYSTATE}.pop
+        printf "File successfully downloaded. Moving to /var/lib/xastir/GNIS\n" 
+        sudo mkdir -p /var/lib/xastir/GNIS
+        sudo mv ${MYSTATE}${SUFFIX} /var/lib/xastir/GNIS/${MYSTATE}.pop
     else
         printf "File for %s not successfully downloaded.\n" ${MYSTATE}
     fi 
diff --git a/scripts/xastir-fixcfg.sh b/scripts/xastir-fixcfg.sh
index 2e6d5e9..1cb637b 100755
--- a/scripts/xastir-fixcfg.sh
+++ b/scripts/xastir-fixcfg.sh
@@ -18,7 +18,7 @@ if [ -f $CNF ]; then
 	echo "$CNF: unable to rename!"
 	exit 1
     fi
-    sed -e 's:/usr/local/xastir/:${prefix}/share/xastir/:' <$CNF.backup >$CNF
+    sed -e 's:/usr/local/xastir/:/var/lib/xastir/:' <$CNF.backup >$CNF
     if [ $? -ne 0 ]; then
 	echo "$CNF: sed failed!"
 	mv $CNF.backup $CNF
diff --git a/scripts/xastir-migrate.sh b/scripts/xastir-migrate.sh
index a15aefb..fef6c11 100755
--- a/scripts/xastir-migrate.sh
+++ b/scripts/xastir-migrate.sh
@@ -9,7 +9,7 @@
 . `dirname $0`/values
 PREFIX=
 OLD=$PREFIX/usr/local/xastir
-NEW=$PREFIX${prefix}/share/xastir
+NEW=$PREFIX/var/lib/xastir
 rr=0
 if [ -d $OLD ]; then
         if [ ! -d $NEW ]; then
diff --git a/src/xa_config.c b/src/xa_config.c
index 085f914..97c7ff0 100644
--- a/src/xa_config.c
+++ b/src/xa_config.c
@@ -429,6 +429,9 @@ char *get_user_base_dir(char *dir, char * base, size_t base_size) {
 char *get_data_base_dir(char *dir) {
     static char base[MAX_VALUE];
     char *env_ptr;
+    struct stat statbuf;
+    char *paths[3];
+    int path;
 
     // Snag this variable from the environment if it exists there,
     // else grab it from the define from the compile command-line
@@ -437,15 +440,31 @@ char *get_data_base_dir(char *dir) {
     // -DXASTIR_DATA_BASE=\"/opt/Xastir/share/xastir\"
     // -DXASTIR_DATA_BASE=\"/usr/local/share/xastir\"
     //
-    xastir_snprintf(base,
-        sizeof(base),
-        "%s",
-        ((env_ptr = getenv ("XASTIR_DATA_BASE")) != NULL) ? env_ptr : XASTIR_DATA_BASE);
+    paths[0] = getenv ("XASTIR_DATA_BASE");
+    paths[1] = XASTIR_STATELESS_BASE;
+    paths[2] = XASTIR_DATA_BASE;
 
-    if (base[strlen (base) - 1] != '/')
-        strncat(base, "/", sizeof(base) - 1 - strlen(base));
+    // Test paths for existing subdir
+    for (path = 0; path < sizeof(paths)/sizeof(paths[0]); path++) {
+        if (paths[path] == NULL)
+            continue;
+
+        xastir_snprintf(base,
+            sizeof(base),
+            "%s",
+            paths[path]);
+
+        if (base[strlen (base) - 1] != '/')
+            strncat(base, "/", sizeof(base) - 1 - strlen(base));
+
+        strncat(base, dir, sizeof(base) - 1 - strlen(base));
+
+        if (stat(base, &statbuf) == 0) {
+            // Just need to know it exists
+            break;
+        }
+    }
 
-    strncat(base, dir, sizeof(base) - 1 - strlen(base));
     return base;
 }
 
-- 
2.18.0

